cylc/rose/utilities.py

```diff
@@ -0,0 +1,441 @@
+        ...     opt_conf_keys='A B',
+        ...     defines=["[env]FOO=BAR"],
+        ...     define_suites=["QUX=BAZ"]
+        ... )
+        >>> node = get_cli_opts_node(opts)
+        >>> node['opts']
+        {'value': 'A B', 'state': '!', 'comments': []}
+        >>> node['env']['FOO']
+        {'value': 'BAR', 'state': '', 'comments': []}
+        >>> node['jinja2:suite.rc']['QUX']
+        {'value': 'BAZ', 'state': '', 'comments': []}
+    """
+    # Unpack options:
+    opt_conf_keys = []
+    defines = []
+    suite_defines = []
+    if opts and 'opt_conf_keys' in dir(opts):
+        opt_conf_keys = opts.opt_conf_keys
+    if opts and 'defines' in dir(opts):
+        defines = opts.defines
+    if opts and 'define_suites' in dir(opts):
+        suite_defines = opts.define_suites
+
+    # Construct new ouput based on optional Configs:
+    newconfig = []
+    newconfig = ConfigNode()
+
+    # For each define determine whether it is an env or template define.
+    for define in defines:
+        match = re.match(
+                (
+                        r'^\[(?P<key1>.*)\](?P<state>!{0,2})'
+                        r'(?P<key2>.*)\s*=\s*(?P<value>.*)'
+                ),
+                define
+            ).groupdict()
+        if match['key1'] == '' and match['state'] in ['!', '!!']:
+            LOG.warning(
+                'CLI opts set to ignored or trigger-ignored will be ignored.'
+            )
+        else:
+            newconfig.set(
+                keys=[match['key1'], match['key2']],
+                value=match['value'],
+                state=match['state']
+            )
+
+    # Write suite defines
+    for define in suite_defines:
+        # For now just assuming that we just support Jinja2 - after I've
+        # Implemented the fully template-engine neutral template variables
+        # section this should be a moot point.
+        match = re.match(
+            r'(?P<state>!{0,2})(?P<key>.*)\s*=\s*(?P<value>.*)', define
+        ).groupdict()
+        newconfig.set(
+            keys=['jinja2:suite.rc', match['key']],
+            value=match['value'],
+            state=match['state']
+        )
+
+    # Specialised treatement of optional configs.
+    if 'opts' not in newconfig:
+        newconfig['opts'] = ConfigNode()
+        newconfig['opts'].value = ''
+    newconfig['opts'].value = merge_opts(newconfig, opt_conf_keys)
+    newconfig['opts'].state = '!'
+
+    return newconfig
+
+
+def add_cylc_install_to_rose_conf_node_opts(rose_conf, cli_conf):
+    if 'opts' in cli_conf:
+        cli_opts = cli_conf['opts'].value
+    else:
+        cli_opts = ''
+    if 'opts' not in rose_conf:
+        rose_conf.set(['opts'], '')
+    rose_conf['opts'].comments = [(
+        f' Config Options \'{cli_opts} (cylc-install)\' from CLI appended to '
+        'options already in `rose-suite.conf`.'
+    )]
+    rose_conf_opts_str = ' '.join(
+        f'{rose_conf["opts"].value} {cli_opts} (cylc-install)'.split()
+    )
```

Comment:
Just spotted, if the `rose-suite.conf` has configured `opts` as `!` "ignored" AND they have provided opt configs on the CLI then I think we just ignore the `!` "ignored" value and use the CLI opts i.e:

```suggestion
    if rose_conf['opts'].state in ['!', '!!']:
        rose_conf_opts_str = cli_opts.split() + ['(cylc-install)']
    else:
        rose_conf_opts_str = ' '.join(
            f'{rose_conf["opts"].value} {cli_opts} (cylc-install)'.split()
        )
```